Routes for creating and joining chats are working

EliSchillinger opened the pull request:
The POST and GET routes for creating and joining chats works based on chat PINS, none of the other routes for the chat page(actually creating a chat/poll) are working yet. Also added the Poll menu but doesn't actually create a poll yet.

EliSchillinger committed:
Routes for creating and joining chats are working

BaoNguyen09 commented on backend/models/Chat.js:
```diff
@@ -2,17 +2,20 @@ const mongoose = require("mongoose");
+		required: true,
+		unique: true,
+	},
+	users: {
```
Why do you remove the validation part?

EliSchillinger replied:
I'll add the validation back in when we have some sort of guest user handling

BaoNguyen09 replied:
Why we need that now? A room must have an user even if it's just guest

EliSchillinger replied:
Because we dont have a user_id for the guest user if they don't exist, so it will just crash every time.

BaoNguyen09 replied:
I don't really get what you mean, every user we create will have an associate _id field with them in the database

Claire-Vu replied:
You can create a guest user using User.create({})

BaoNguyen09 commented on backend/routes/chatRouter.js:
```diff
@@ -1,6 +1,69 @@
 const chatRouter = express.Router();
+const Chat = require("../models/Chat.js");
+
+// Join a chat
```
This one belongs to other issue: #18

BaoNguyen09 commented on backend/routes/chatRouter.js:
```diff
@@ -1,6 +1,69 @@
+    }
+});
+
+//Utility function to generate PIN
```
This one needs discussion since right now we're using the chat id generated by MongoDB when create new Chat documents 
What do you guys think? @matthewsong18 @Claire-Vu

BaoNguyen09 commented on backend/routes/chatRouter.js:
We put all functions that interact between routes and database into controllers folder

EliSchillinger replied:
Why are we doing this? Seems like unnecessary complexity to me.

BaoNguyen09 replied:
Because we've had too much code to just put all into the router file and it's easier to avoid merge conflicts when working together

Claire-Vu replied:
It should be pretty easy to do, just move the async function into the chatController folder and then call the function like with the routes in userRouter

EliSchillinger committed:
Created a poll pop-up, fixed bug where the rout to get a poll and the route to join a chat were the same

EliSchillinger committed:
Made it so that options can be added to the poll menu

Claire-Vu commented on backend/routes/chatRouter.js:
```diff
@@ -1,33 +1,80 @@
+		//Check if a chat with the given PIN exists
+		console.log("Attempting to find chat");
+        const chat = await Chat.findOne({ pin: chatId });
+		console.log("Chat found");
```
This would log to the console even if it was not found

BaoNguyen09 replied:
solved

Claire-Vu commented on backend/routes/chatRouter.js:
```diff
@@ -1,33 +1,80 @@
+chatRouter.post("/create", async (req, res) => {
+	console.log("POST request recieved");
+    const pin = generateRandomPin();
+    const chatName = "Anonymous Chat";
```
Do we need a chat name if everything is going to just be named Anonymous Chat?

BaoNguyen09 replied:
For now, i leave that field as unrequired so that our current code does not fail but we implement or remove it later

Claire-Vu commented on backend/routes/chatRouter.js:
```diff
@@ -1,33 +1,80 @@
+        const chat = await Chat.findOne({ pin: chatId });
+		console.log("Chat found");
+
+        if (chat) {
+			//Respond true if the chatroom exists
+			res.json({ exists: true });
```
I think here we also need to add the joined user into the chat Users field and the chat to the chats field of the user

BaoNguyen09 replied:
Since that this point, we still don't have a way to determine who that user is (guest or not), i will let the `chat/:chat_id` handles this (it will determine if that user is guest and create new User document accordingly and associate the chat to it)

BaoNguyen09 replied:
For now i just handle this for guest user, but when we use cookie we can easily determine that user's identify

Claire-Vu commented on backend/routes/chatRouter.js:
```diff
@@ -1,33 +1,80 @@
+        const newChat = new Chat({
+            name: chatName,
+            pin: pin,
+            users: [],
```
You can probably create the guest user by doing what I stated above and add the guest here now

BaoNguyen09 replied:
Solved

Claire-Vu commented on frontend/public/styles/chat.css:
I found an issue with the poll overflowing if we add too many options

BaoNguyen09 replied:
So once this PR is merged, @EliSchillinger can make a branch on issue #25 to continue working on it

Claire-Vu reviewed (commented):
Great work, just some minor changes needed.

BaoNguyen09 committed:
Merge remote-tracking branch 'origin/dev' into 17-Chat-Router

BaoNguyen09 committed:
fix: add validation back for ChatSchema

BaoNguyen09 committed:
refactor: move code handling request to controller

I also rearrange the order of each function in chatRouter to original one

BaoNguyen09 commented on backend/models/Chat.js:
```diff
@@ -2,17 +2,20 @@ const mongoose = require("mongoose");
-  },
-  message: [{ type: Schema.Types.ObjectId, ref: "Message" }],
-  polls: [{ type: Schema.Types.ObjectId, ref: "Poll" }],
+	name: {
+		type: String,
+		required: true,
+	},
```
I don't think we need a name for each chat since they will just be anoynomous anyway

BaoNguyen09 committed:
refactor: rename this route following the structure of other routes

BaoNguyen09 committed:
fix: rename this route to avoid conflict with join route

BaoNguyen09 committed:
refactor: modify join_chat to correctly validate chat pin

I also modify respond_with_error_json function to be more flexible with different reponse json and status code

BaoNguyen09 committed:
fix: ensure pin generated is unique to avoid chat collision

BaoNguyen09 committed:
fix: remove duplicate imports of create_message

BaoNguyen09 committed:
fix: ensure that each new chat has at least 1 user

I also wrap code into try-catch block to capture all errors

BaoNguyen09 committed:
chore: update old routes

BaoNguyen09 committed:
Merge remote-tracking branch 'origin/17-Chat-Router' into 17-Chat-Router

BaoNguyen09 committed:
fix: remove /api route since we don't have it now

BaoNguyen09 commented on backend/routes/chatRouter.js:
```diff
@@ -12,25 +13,15 @@ chatRouter.get("/:chat_id", (req, res) => {
-    res.send(`Create a new chat`);
-});
-
-// Handle undefined routes
-chatRouter.get("*", (req, res) => {
-    res.sendFile(path.join(__dirname, "../../frontend/public/error.html"))
-})
```
This code is moved down below (not deleted) by @EliSchillinger

BaoNguyen09 committed:
fix: unrequire these new fields to prevent current code from failing

BaoNguyen09 reviewed (commented):
Some issues remaining:

1. `/chat/:chat_id/join` route needs to be modified when we use cookie to store user_id
2. `/chat/create` route needs to be modified when we use cookie to store user_id
3. Poll form will be overflown if there're too many poll options created (frontend)

Claire-Vu commented on backend/controllers/chatController.js:
```diff
@@ -235,12 +233,104 @@ function stringtify_for_sse(raw_data) {
   return `data: ${JSON.stringify(raw_data)}\n\n`;
 }
 
+// TO join a chat:
+// 1. Validate chat_pin.
+// 2. Repond with exists: true if found
+// 3. Respond with error and exists: false if not.
+async function join_chat(req, res) {
+	console.log("GET request received");
+  const chat_pin = req.params.chat_id;
+  try {
+    console.log("Attempting to find chat");
+    validate_chat_pin(chat_pin);
+    console.log("Chat found");
+    res.status(200).json({ exists: true });
+  } catch (error) {
+    console.error("Error checking chatroom existence:", error);
+    respond_with_error_json(res, 400, {exists: false, message: error.message});
+  }
+}
```
need to add to user chats too when they join a chat

BaoNguyen09 replied:
For now I may just handle it for guest user, but once we have cookie implmented, we can then handle for users with account. What do you think?

Claire-Vu replied:
Alright, if we didn't wanna do cookies we could also just have the user_id stored inside the HTML

BaoNguyen09 replied:
But then whichever page we redirect a user to, we will have to send that user_id around which is quite a lot of work for client-side Js. If we use cookie, we can just simply assign user_id for the browser to keep and can even simplify the post message route (since we can just extract the user_id from the cookie, instead of passing the user_name around)

BaoNguyen09 committed:
fix: link guest users and chat together
